File: 145 Binary Tree Postorder Traversal/145.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // Solution 1: reverse
    public List<Integer> postorderTraversal(TreeNode root) {
        // use the pre-order, then reverse
        // you can use two stack or Collections.reverse(result))
        // or append these items from the reverse order!
        List<Integer> res = new LinkedList<>();
        if (root == null) {
            return res;
        }
        LinkedList<TreeNode> stack = new LinkedList<>();
        TreeNode cur = root;
        stack.offerFirst(root);
        while (!stack.isEmpty()) {
            cur = stack.pollFirst();
            // this order of cur.left and cur.right also need to change
            // pre-order is right > left
            // post-order id left => right
            if (cur.left != null) {
                stack.offerFirst(cur.left);
            }
            if (cur.right != null) {
                stack.offerFirst(cur.right);
            }
            res.add(0, cur.val); // so amazing, isn't it ?
        }
        return res;
        // What is the drawback of Solution 1 if we print the nodes on the fly?
        // Need to store every node in memory before we can get the whole post order traversal sequence. 
        // Not real time print
    }

    public List<Integer> postorderTraversal2(TreeNode root) {
        if (root == null) {
            return root;
        }
        List<Integer> res = new LinkedList<>();
        TreeNode cur = root, pre = null;
        Deque<TreeNode> stack = new LinkedList<>();
        stack.offerFirst(root);
        while (!stack.isEmpty()) {
            // cur = stack.pollFirst(); no!!!!
            cur = stack.peekFirst();
            if (pre == null || pre.left == cur || pre.right == cur) {
                if (cur.left != null) {
                    stack.offerFirst(cur.left);
                } else if (cur.right != null) {
                    stack.offerFirst(cur.right);
                } else {
                    res.add(cur.val);
                    stack.pollFirst();
                }
            } else if (pre == cur.left) {
                if (cur.right != null) {
                    stack.offerFirst(cur.right);
                } else {
                    res.add(cur.val);
                    stack.pollFirst();
                }
            } else {
                res.add(cur.val);
                stack.pollFirst();
            }
            pre = cur;
        }
        return res;
    }
}

File: 145 Binary Tree Postorder Traversal/145.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        if not root:
            return res
        stack = deque()
        stack.append(root)
        cur = root
        while stack:
            cur = stack.pop()
            if cur.left:
                stack.append(cur.left)
            if cur.right:
                stack.append(cur.right)
            res.insert(0, cur.val)
        return res

    def postorderTraversal2(self, root: TreeNode) -> List[int]:
        res = []
        if not root:
            return res
        stack = deque()
        stack.append(root)
        cur = root
        pre = None
        while stack:
            cur = stack[-1]
            if not pre or pre.left == cur or pre.right == cur:
                if cur.left:
                    stack.append(cur.left)
                elif cur.right:
                    stack.append(cur.right)
                else:
                    res.append(cur.val)
                    stack.pop()
            elif pre == cur.left:
                if cur.right:
                    stack.append(cur.right)
                else:
                    res.append(cur.val)
                    stack.pop()
            else:
                res.append(cur.val)
                stack.pop()
            pre = cur
        return res
